### PR TITLE
[binutils] Bump zlib version

### PR DESCRIPTION
**Pull request by samuel-emrys**
* Bump to zlib/1.2.13 to support #13896

Specify library name and version:  **binutils/2.38**

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by uilianries** [comment]
@samuel-emrys thank you for updating it too!

**Comment by uilianries** [comment]
@samuel-emrys The CI linter requests some imports to be updated. Could you please updated them according to the list https://github.com/conan-io/conan-center-index/blob/master/docs/v2_linter.md#import-tools-from-conan

File: recipes/binutils/all/conanfile.py [hunk 1]
```diff
@@ -56,8 +59,8 @@ def _settings_target(self):
         return getattr(self, "settings_target", None) or self.settings
 
     def export_sources(self):
-        for patch in self.conan_data.get("patches", {}).get(self.version, []):
-            self.copy(patch["patch_file"])
+        for _patch in self.conan_data.get("patches", {}).get(self.version, []):
+            self.copy(_patch["patch_file"])
```
Comment (prince-chrismc) [inline recipes/binutils/all/conanfile.py:63]: There's an export patches helper from 1.53 you can add here

File: recipes/binutils/all/conanfile.py [hunk 2]
```diff
@@ -126,64 +129,78 @@ def _raise_unsupported_configuration(self, key, value):
...
-        ]
-        autotools.configure(args=conf_args, configure_dir=self._source_subfolder)
-        return autotools
+        tc = AutotoolsToolchain(self)
+        tc.configure_args.append("--disable-nls")
+        tc.configure_args.append(f"--target={self.options.target_triplet}")
+        tc.configure_args.append(f"--enable-multilib={yes_no(self.options.multilib)}")
+        tc.configure_args.append(f"--with-zlib={self.deps_cpp_info['zlib'].rootpath}")
+        tc.configure_args.append(f"--program-prefix={self.options.prefix}")
+        tc.configure_args.append(f"--exec_prefix={unix_path(self, os.path.join('/','bin', 'exec_prefix'))}")
```
Comment (prince-chrismc) [inline recipes/binutils/all/conanfile.py:154]: This new syntax 🤢 

This broke my brain for a good 30 seconds -- https://docs.conan.io/en/latest/reference/conanfile/tools/microsoft.html?highlight=unix_path

***

Why does this need to be root? should it not be relative? the old one was absolute to the package folder https://github.com/conan-io/conan-center-index/pull/14137/files#diff-c68fc5112c8ad8e228d2ea59d42386921fb01b28ab2be48ec5f060a0218a76ffR143

File: recipes/binutils/all/conanfile.py [hunk 3]
```diff
@@ -126,64 +129,78 @@ def _raise_unsupported_configuration(self, key, value):
...
+        tc.generate()
 
     def build(self):
-        for patch in self.conan_data.get("patches", {}).get(self.version, []):
-            tools.patch(**patch)
-        autotools = self._configure_autotools()
+        for _patch in self.conan_data.get("patches", {}).get(self.version, []):
+            patch(self, **_patch)
```
Comment (prince-chrismc) [inline recipes/binutils/all/conanfile.py:159]: There's a new apply patches in 1.52 that could be used here

File: recipes/binutils/all/conanfile.py [hunk 4]
```diff
@@ -126,64 +129,78 @@ def _raise_unsupported_configuration(self, key, value):
...
+            pattern="COPYING*",
+            dst=os.path.join(self.package_folder, "licenses"),
+            src=self.source_folder,
+            keep_path=False,
+        )
 
     def package_info(self):
+        target_bindir = os.path.join(self._exec_prefix, str(self.options.target_triplet), "bin")
+        self.output.info("Prepending PATH environment variable: {}".format(target_bindir))
+        self.buildenv_info.prepend_path("PATH", target_bindir)
```
Comment (prince-chrismc) [inline recipes/binutils/all/conanfile.py:181]: Not required - it's added by default in the new generator
```suggestion
```

File: recipes/binutils/all/conanfile.py [hunk 5]
```diff
@@ -126,64 +129,78 @@ def _raise_unsupported_configuration(self, key, value):
...
+
+        binaries = os.listdir(target_bindir)
+        self.output.info(f"Binaries built: {', '.join(binaries)}")
+        for binary_name in binaries:
+            binary = os.path.join(target_bindir, binary_name)
+            if os.path.isfile(binary):
+                self.output.info(f"Setting {binary_name.upper()}={binary}")
+                self.buildenv_info.define(f"{binary_name.upper()}", binary)
```
Comment (prince-chrismc) [inline recipes/binutils/all/conanfile.py:189]: What is this exporting exacly?

IT's better to pt a more explicit list - this way we can review and understand the implaations. 🙏

**Review by prince-chrismc — COMMENTED** [review]
2 mall suggestions and a few questions

File: recipes/binutils/all/conanfile.py (hunk 2 above)
Comment (samuel-emrys) [reply recipes/binutils/all/conanfile.py:154]: I'm not convinced that this was doing anything in the previous version, note the missing `--` prepended to `exec_prefix`. The confusing thing about this line, and the way gnu projects seem to accept paths in general, is that this _is_ a relative path, even though it looks like an absolute path with the leading `/`. If this is set to the `package_folder`, then the `exec_prefix` ends up being something like `/home/user/.conan/data/binutils/2.38/_/_/package/<package id>/`, which gets embedded in the exported package - not ideal.

Long story short - works as intended. I guess given that this would only ever run in a unix environment, even on windows, I guess this could be simplified to
```python
        tc.configure_args.append(f"--exec_prefix=/bin/exec_prefix")
```

I was concerned that path separators could make things awkward on windows, but yeah given that this is a relative path and can only ever be built in a bash environment this is probably safe enough and much more readable.

File: recipes/binutils/all/conanfile.py (hunk 4 above)
Comment (samuel-emrys) [reply recipes/binutils/all/conanfile.py:181]: This isn't the `bin` directory in the package folder. Relative to the package folder, this would be `{package_folder}/bin/exec_prefix/x86_64-pc-linux-gnu/bin`, so not automatically added to the path to my understanding.

File: recipes/binutils/all/conanfile.py (hunk 5 above)
Comment (samuel-emrys) [reply recipes/binutils/all/conanfile.py:189]: the binaries include `["ar", "nm", "objcopy", "objdump", "ranlib", "readelf", "strip", "as", "ld"]` on linux. Which means that this just exports those names in upper case, i.e:
* `AR={package_folder}/bin/exec_prefix/x86_64-pc-linux-gnu/bin/ar`
* `NM={package_folder}/bin/exec_prefix/x86_64-pc-linux-gnu/bin/nm`
* `OBJCOPY={package_folder}/bin/exec_prefix/x86_64-pc-linux-gnu/bin/objcopy`

The reason I chose to resolve this programmatically is because the collection of binaries built is reasonably dependent on the platform and architecture it is building for. To see the logic that it uses to choose what targets are built, see:

https://gitlab.com/git-mirrors/binutils/-/blob/master/configure#L3804-4103

To illustrate with an example, on x86_64 macOS, `binutils` will build everything in the above list with the exception of the assembler (`as`) and linker (`ld`) because these aren't supported on that platform. Anyway, the logic here simplifies the export logic reasonably substantially by exporting all binaries that were built with a variable name to match the uppercase name of the target.

I think that this is pretty readable - if `binary_name.upper()` isn't self documenting, I'm not sure what is. I think that this is the best way to deal with this though, I can add a comment to clarify the binaries that would commonly be included and an example of the environment variable if you'd like? These are also printed to the console, so it would be reasonably clear to the user.

File: recipes/binutils/all/conanfile.py (hunk 4 above)
Comment (SpaceIm) [reply recipes/binutils/all/conanfile.py:179]: you should add this path to cpp_info.bindirs

File: recipes/binutils/all/conanfile.py [hunk 6]
```diff
@@ -1,6 +1,6 @@
 from conan import ConanFile
 from conan.tools.gnu import Autotools, AutotoolsToolchain
-from conan.tools.files import get, rmdir, rm, copy, patch
+from conan.tools.files import get, rmdir, rm, copy, apply_conandata_patches, export_conandata_patches
 from conan.errors import ConanInvalidConfiguration
 from conan.tools.microsoft import is_msvc, unix_path
```
Comment (prince-chrismc) [inline recipes/binutils/all/conanfile.py:5]: Linter

```suggestion
from conan.tools.microsoft import is_msvc
```

File: recipes/binutils/all/conanfile.py [hunk 7]
```diff
@@ -126,64 +128,77 @@ def _raise_unsupported_configuration(self, key, value):
...
+        self.cpp_info.bindirs = ["bin", target_bindir]
+
+        absolute_target_bindir = os.path.join(self.package_folder, target_bindir)
+        binaries = os.listdir(absolute_target_bindir)
+        self.output.info(f"Binaries built: {', '.join(binaries)}")
+        for binary_name in binaries:
+            binary = os.path.join(absolute_target_bindir, binary_name)
+            if os.path.isfile(binary):
+                self.output.info(f"Setting {binary_name.upper()}={binary}")
+                self.buildenv_info.define(f"{binary_name.upper()}", binary)
```
Comment (prince-chrismc) [inline recipes/binutils/all/conanfile.py:187]: Thanks for filling me in -- let just leave a trace for future reviewers so it makes sense

```suggestion
                # See https://github.com/conan-io/conan-center-index/pull/14137/files/7ed7a48e2c993bb7e748570ee7ab9d021790c7dc..e93ea8d3d318b0ef8be84816cef615043b5f7fa9#r1022730130 for details
                self.output.info(f"Setting {binary_name.upper()}={binary}")
                self.buildenv_info.define(f"{binary_name.upper()}", binary)
```

**Comment by prince-chrismc** [comment]
3 tiny suggestions for you then I'll give it a check! 

great work

**Review by prince-chrismc — APPROVED** [review]

**Review by uilianries — APPROVED** [review]
LGTM

File: recipes/binutils/all/conanfile.py (hunk 5 above)
Comment (jcar87) [reply recipes/binutils/all/conanfile.py:188]: I've landed here from the comment left in the recipe :) 

I think that reading the recipe alone it's not entirely clear (as a reader), what this is doing fully, in order to work this out, one must remember which executables are part of the binutils package, and there's several.

The secondary thing is that as a reviewer, it's also not clear _why_ these variables need to be defined and why they need to be absolute paths - can't downstream consumers not rely on `PATH` being defined during `./configure` to find the correct tools inside _this_ package?

I think don't think there are problems with readability per se, but perhaps if there had been a explicit list, we could have collectively identified this https://github.com/conan-io/conan-center-index/issues/14566 before it happened.
